aptos-move/aptos-aggregator/src/aggregator_change_set.rs

```diff
@@ -2,331 +2,388 @@
-                )),
-                AggregatorChange::AggregatorDelta {
-                    delta: prev_delta,
-                    max_value: prev_max_value,
-                    history: prev_history,
-                } => {
-                    if max_value != prev_max_value {
-                        return Err(code_invariant_error(
-                            "Max value of aggregator cannot change",
-                        ));
-                    }
-                    let (new_delta, new_history) =
-                        merge_two_deltas(prev_delta, prev_history, delta, history, *max_value)?;
-                    *self = AggregatorChange::AggregatorDelta {
-                        delta: new_delta,
-                        history: new_history,
-                        max_value: *max_value,
-                    };
-                    Ok(())
-                },
-                AggregatorChange::SnapshotDelta { .. } => Err(code_invariant_error(
-                    "SnapshotDelta cannot come before AggregatorDelta",
-                )),
+    pub fn get_apply_base_id(&self, self_id: &AggregatorID) -> ApplyBase {
+        self.get_apply_base_id_option()
+            .unwrap_or(ApplyBase::Previous(*self_id))
+    }
+
+    pub fn apply_to_base(&self, base_value: AggregatorValue) -> PartialVMResult<AggregatorValue> {
+        use AggregatorApplyChange::*;
+
+        Ok(match self {
+            AggregatorDelta { delta } => {
+                AggregatorValue::Aggregator(delta.apply_to(base_value.into_aggregator_value()?)?)
+            },
+            SnapshotDelta { delta, .. } => {
+                AggregatorValue::Snapshot(delta.apply_to(base_value.into_aggregator_value()?)?)
             },
-            AggregatorChange::SnapshotDelta { .. } => {
-                unimplemented!()
+            SnapshotDerived { formula, .. } => {
+                AggregatorValue::Derived(formula.apply(base_value.into_snapshot_value()?))
             },
+        })
+    }
+}
+
+impl AggregatorChange {
+    pub fn get_merge_dependent_id(&self) -> Option<AggregatorID> {
+        use AggregatorApplyChange::*;
+        use AggregatorChange::*;
+
+        match self {
+            // Only SnapshotDelta merging logic depends on current aggregator change
+            Apply(SnapshotDelta {
+                base_aggregator, ..
+            }) => Some(*base_aggregator),
+            Create(_) | Apply(AggregatorDelta { .. } | SnapshotDerived { .. }) => None,
         }
     }
 
-    /// Applies next aggregator change on top of self, merging two changes together. This is a reverse
-    /// of `merge_with_previous_aggregator_change`.
-    pub fn merge_with_next_aggregator_change(
-        &mut self,
+    /// Applies next aggregator change on top of self, merging two changes together.
+    pub fn merge_two_changes(
+        prev_change: Option<&AggregatorChange>,
+        prev_dependent_change: Option<&AggregatorChange>,
```

Comment:
What is prev_change and prev_dependent_change? Having a comment would help.

Reply:
will add a comment. 

prev_change is change from the same AggregatorID as next_change
prev_dependent_change is change from the different AggregatorID - i.e. aggregator that snapshot depends on

Reply:
In the above example, 
change1: agg1 -> Delta(+3)
change2: agg1 -> Delta(+6), snap1 -> (base=agg1, Delta(+2))

Here,    change1(agg2) is prev_dependent_change.   change2(snap1) is prev_change.  What is next_change then?
I didn't understand why "merge_two_changes" function takes 3 changes as input.